### ISSUE TITLE
decodeList discards error messages from list items

### ISSUE THREAD

**Issue by davidbkemp** [circe/circe#596]
When decoding a list, if any of the items in the list fail to decode, then the failure message from `Decoder. decodeList` is always `"[A]List[A]"` and does not include any of the failure messages from the individual items.  This makes trouble shooting difficult.

The relevant line is:
https://github.com/circe/circe/blob/master/modules/core/shared/src/main/scala/io/circe/Decoder.scala#L664

The same applies to `decodeSet` and `decodeVector`.

**Comment by kevinmeredith** [comment]
Could you please provide an example, @davidbkemp? I'm curious of the code demonstrating the issue that you've raised here.

**Comment by kevinmeredith** [comment]
Actually, Travis helped me [out]().

You're talking about this scenario, @davidbkemp?

```
import io.circe._
import parser._
import syntax._

scala> val ints = parse("[1,2]")
ints: Either[io.circe.ParsingFailure,io.circe.Json] =
Right([
  1,
  2
])

scala> Decoder.decodeList[Int].decodeJson(ints.right.get)
res0: io.circe.Decoder.Result[List[Int]] = Right(List(1, 2))

scala> val strings = parse("[\"a\"]")
strings: Either[io.circe.ParsingFailure,io.circe.Json] =
Right([
  "a"
])

scala> Decoder.decodeList[Int].decodeJson(strings.right.get)
res2: io.circe.Decoder.Result[List[Int]] = Left(DecodingFailure([A]List[A], List(DownArray)))
```

Per this issue, you'd prefer a better `DecodingFailure` "message" (if that's the right word) than `[A]List[A]`?

**Comment by davidbkemp** [comment]
Yes, it would be nice if the "message" included the decoding failure message for attempting to decode the string as an integer. The current message "[A]List[A]" doesn't tell you much, especially if the actual failure occurs deep inside the attempt to decode one of the A's.

**Comment by lukestephenson** [comment]
@cponomaryov Thanks for fixing this. 

Looking forward to a new release with this fix.

**Comment by travisbrown** [comment]
Resolved in #649.
